api/src/main/java/org/apache/iceberg/expressions/AggregateEvaluator.java

```diff
@@ -0,0 +1,118 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.expressions;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+
+public class AggregateEvaluator {
+
+  public static AggregateEvaluator create(List<BoundAggregate<?, ?>> aggregates) {
+    return new AggregateEvaluator(aggregates);
+  }
+
+  public static AggregateEvaluator create(Schema schema, List<Expression> aggregates) {
+    return create(schema.asStruct(), aggregates);
+  }
+
+  private static AggregateEvaluator create(Types.StructType struct, List<Expression> aggregates) {
+    List<BoundAggregate<?, ?>> boundAggregates =
+        aggregates.stream()
+            .map(expr -> Binder.bind(struct, expr))
+            .map(bound -> (BoundAggregate<?, ?>) bound)
+            .collect(Collectors.toList());
+
+    return new AggregateEvaluator(boundAggregates);
+  }
+
+  private final List<BoundAggregate.Aggregator<?>> aggregators;
+  private final Types.StructType resultType;
+  private final List<BoundAggregate<?, ?>> aggregates;
+
+  private AggregateEvaluator(List<BoundAggregate<?, ?>> aggregates) {
+    ImmutableList.Builder<BoundAggregate.Aggregator<?>> aggregatorsBuilder =
+        ImmutableList.builder();
+    List<Types.NestedField> resultFields = Lists.newArrayList();
+    for (int pos = 0; pos < aggregates.size(); pos += 1) {
+      BoundAggregate<?, ?> aggregate = aggregates.get(pos);
+      aggregatorsBuilder.add(aggregate.newAggregator());
+      resultFields.add(Types.NestedField.optional(pos, aggregate.describe(), aggregate.type()));
+    }
+
+    this.aggregators = aggregatorsBuilder.build();
+    this.resultType = Types.StructType.of(resultFields);
+    this.aggregates = aggregates;
+  }
+
+  public void update(StructLike struct) {
+    for (BoundAggregate.Aggregator<?> aggregator : aggregators) {
+      aggregator.update(struct);
+    }
+  }
+
+  public void update(DataFile file) {
+    for (BoundAggregate.Aggregator<?> aggregator : aggregators) {
+      aggregator.update(file);
+    }
+  }
+
+  public Types.StructType resultType() {
+    return resultType;
+  }
+
+  public StructLike resultStruct() {
```

Comment:
I think this is  for future usage @rdblue

Reply:
This should be how data is returned by the evaluator, right? How would Spark get the result if not by calling this method?

Reply:
Oh, I see. My original PR had `public StructLike result()` rather than two methods. I think that was correct. Everything should get the result as `StructLike` and wrap to adapt to other in-memory object models.

Reply:
Sorry I initially didn't know this `StructInternalRow` wrapper class, so I broke your method into two and used `Object[]` directly, and then I forgot I have changed your original method. I changed it back.